### ISSUE TITLE
Alternative specification for pipe formatting

### ISSUE THREAD

**Issue by BigCrow** [nushell/nufmt#60]
The [specification.md](https://github.com/nushell/nufmt/blob/main/docs/specification.md#sensible--default-and-features) describes the default formatting for pipelines to be a wall of lines.
```nu
def "apply to" [
    file: path
    modification: closure
] {
    $file
    | path expand
    | open --raw
    | from toml
    | do $modification
    | save --force $file
}
```
In the nushell discord [fdncred](https://github.com/fdncred) mentioned how this creates issues with copy pasting code to the terminal on windows. Because of this I would like to suggest alternative formatting & indenting:
```nu
def "apply to" [
    file: path
    modification: closure
] {
    $file |
        path expand |
        open --raw |
        from toml |
        do $modification |
        save --force $file
}
```
This python style indenting still showcases that a pipeline is continued, despite that the pipe symbol is not at the start of the line.

**Comment by fdncred** [comment]
Agreed.

**Comment by amtoine** [comment]
i think this should be modifiable in some settings, i personally quite strongly dislike the second way :rofl:

**Comment by CabalCrow** [comment]
> i think this should be modifiable in some settings, i personally quite strongly dislike the second way 🤣
> 
> and i would write the following instead
> 
> ```nushell
> def "apply to" [
>     file: path
>     modification: closure
> ] {
>     $file
>         | path expand
>         | open --raw
>         | from toml
>         | do $modification
>         | save --force $file
> }
> ```

This falls into the same issue for windows users - you can't have a | at the start of the line, indent or not.

**Comment by amtoine** [comment]
@CabalCrow 

'tis why i propose to make in configurable :)

**Comment by CabalCrow** [comment]
@amtoine Having it configurable, is good, but I'm also talking about what the default formatting should be. I believe that you would want default formatting that could work properly regardless of OS, since nushell aims to work regardless of OS.

**Comment by AucaCoyan** [comment]
> This falls into the same issue for windows users - you can't have a | at the start of the line, indent or not.

What do you mean with Windows users? I can copy and paste that script in my nushell terminal and it works 🤩. Do you have trouble with that script?

**Comment by amtoine** [comment]
yeah i think Nushell should be able to handle copy pasting any way to format pipelines, the parser shouldn't care at all :confused: 

if that's not possible, i agree the default should work the best for everyone -even though it's Windows which is annoying again, ugh...-, i would just turn that default off as soon as i start a Nushell project :wink: 

my gut feeling is that, putting the `|` at the end of lines would like doing the following in Rust
```rust
            let vals = rec.
                values().
                cloned().
                enumerate().
                map(|(i, v)| {
                    if i == id {
                        mutate_value_cell(&v, &cell_path, cell).unwrap()
                    } else {
                        v
                    }
                }).
                collect();
```
instead of the conventional
```rust
            let vals = rec
                .values()
                .cloned()
                .enumerate()
                .map(|(i, v)| {
                    if i == id {
                        mutate_value_cell(&v, &cell_path, cell).unwrap()
                    } else {
                        v
                    }
                })
                .collect();
```

anyways :wink:

**Comment by amtoine** [comment]
@AucaCoyan
if there is no issue anymore with copy pasting pipelines on Windows, then it's all good :relieved:

**Comment by CabalCrow** [comment]
On 24/06/02 04:00AM, Antoine Stevan wrote:
> @AucaCoyan
> if there is no issue anymore with copy pasting pipelines on Windows, then it's all good :relieved: 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/nushell/nufmt/issues/60#issuecomment-2143802164
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

[fdncred](https://github.com/fdncred) shared the issue. I'm personally
not on windows so I can't talk about the issue first hand.

**Comment by fdncred** [comment]
Here is an example @AucaCoyan, if your terminal does not support bracketed paste, such as Windows Terminal, when you paste the middle portion of the custom command, something like this
```nushell
    $file
        | path expand
        | open --raw
        | from toml
        | do $modification
        | save --force $file
```
nushell/reedline pastes one line at a time. So, it looks like this to nushell and each line tries to be executed giving you a different result than what you want.
```nushell
$file # it runs this
```
```nushell
    | path expand # then tries to run this
```
```nushell
    | open --raw # then tries to run this, and so on
```
alternatively, if you paste this
```nushell
    $file |
        path expand |
        open --raw |
        from toml |
        do $modification |
        save --force $file
```
nushell sees
```nushell
$file | # oh, this statement isn't complete, don't execute it yet
```
```nushell
    path expand | # oh, there's another pipe, so this statement isn't complete, don't execute yet
```
```nushell
   open --raw | # oh, another pipe, don't execute yet
```

So, all I was saying is that I prefer the pipes at the end because Windows users won't have a problem with copying and pasting pieces out of the script/custom command. You are correct that the entire custom command can be pasted fine but that's because it's in a `{` `}` block but if you try to play with just the middle pieces, it won't work.

You can see here that I pasted the entire custom command and it worked just fine. But if I want just the part in question, it pukes because the pipes are in the wrong place for windows to understand what to do.
![image](https://github.com/nushell/nufmt/assets/343840/4c58e676-afc1-4e16-be89-d9187b4118b6)
But if the pipes are at the end, windows has no issues with it
![image](https://github.com/nushell/nufmt/assets/343840/f016d889-dfb2-492d-b919-57924ed67443)

**Comment by AucaCoyan** [comment]
Thanks for the clarification! I tried also in Windows and had the same result in Windows Terminal. Because I've been using nushell in windows for so long I didn't expect it to "magically work" in linux 😋.
BTW, is there any other terminal in windows that supports bracketed paste? Wezterm and Rio also couldn't resolve the job

**Comment by fdncred** [comment]
I'm not sure. I'm hopeful that once Ghostty gets fully supported in Windows that it will be I'm not sure. I think in most terminals you have to enable the feature though.
